lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1459,37 +1457,39 @@ class HSAQueue final : public KalmarQueue
                     << (op->getCommandKind() == hcCommandKernel ? ((static_cast<HSADispatch*> (op.get()))->getKernelName()) : "")  // change to getLongKernelName() for mangled name
                     << std::endl);
 
-        if (((hsa_signal_t*)op->getNativeHandle())->handle == 0) {
-            // no signal present in the new HSAOp, so add to this queue's vector of signal-less ops
-            asyncOpsWithoutSignal.push_back(op);
-        }
-        else if (!asyncOpsWithoutSignal.empty()) {
-            // signal is present in the new HSAOp, and this queue has one or more signal-less ops needing a signal association
-            DBOUT(DB_SIG, "attaching " << asyncOpsWithoutSignal.size() << " ops without signals to " << *op
-                    << " signal="<< std::hex  << ((hsa_signal_t*)op->getNativeHandle())->handle << std::dec
-                    << " commandKind=" << getHcCommandKindString(op->getCommandKind())
-                    << " "
-                    << (op->getCommandKind() == hcCommandKernel ? ((std::static_pointer_cast<HSADispatch> (op))->getKernelName()) : "")  // change to getLongKernelName() for mangled name
-                    << std::endl);
-            if (!op->asyncOpsWithoutSignal.empty()) {
-                throw Kalmar::runtime_exception("new op does not have empty async ops list!", 0);
+        auto op_getting_replaced = asyncOps[asyncOpsIndex]; // ref counts up, if it was a valid HSAOp
+        if (op_getting_replaced != nullptr) {
+            if (static_cast<hsa_signal_t*>(op_getting_replaced->getNativeHandle())->handle == 0) {
+                // No signal present in the op_getting_replaced, so it won't wait() on it as part of its dtor.
+                // Add it to this queue's vector of signal-less ops to free once we know they're finished.
+                asyncOpsWithoutSignal.push_back(std::move(op_getting_replaced));
```

Comment:
May I suggest to rename `asyncOpsWithoutSignal` to `asyncOpsWithoutSignalWaitingNextSync`?  I think it would help remembering what this is for.